### ISSUE TITLE
salt-minion crashes when network is broken during initialization

### ISSUE THREAD

**Issue by alf** [saltstack/salt#24283]
I have a few minions that seem to have a very flaky zeromq channel to the salt master. I have the following minion configuration to try to work around this:

```yaml
# Detect disconnects from the master and recover when it happens
ping_interval: 3 # in minutes
auth_tries: 5 # times to try to reconnect before restarting the minion
auth_safemode: False # restart the minion if we can't reconnect
restart_on_error: True # Needed on 2014.7, removed in 2015.2
```

The problem I'm seeing is that these minions tend to crash with the following log messages:

```
2015-03-26 12:19:30,180 [salt.minion      ][INFO    ] Returning information for job: 20150326111935085776
2015-03-26 12:24:11,157 [salt.payload     ][INFO    ] SaltReqTimeoutError: after 60 seconds. (Try 1 of 5)
2015-03-26 12:25:11,218 [salt.payload     ][INFO    ] SaltReqTimeoutError: after 60 seconds. (Try 2 of 5)
2015-03-26 12:26:11,279 [salt.payload     ][INFO    ] SaltReqTimeoutError: after 60 seconds. (Try 3 of 5)
2015-03-26 12:27:11,338 [salt.payload     ][INFO    ] SaltReqTimeoutError: after 60 seconds. (Try 4 of 5)
2015-03-26 12:28:11,339 [salt.payload     ][INFO    ] SaltReqTimeoutError: after 60 seconds. (Try 5 of 5)
2015-03-26 12:28:11,340 [salt             ][ERROR   ] Attempt to authenticate with the salt master failed
2015-03-26 12:28:11,340 [salt             ][WARNING ] ** Restarting minion **
2015-03-26 12:28:11,340 [salt             ][INFO    ] Sleeping random_reauth_delay of 9 seconds
2015-03-26 12:29:20,890 [salt.payload     ][INFO    ] SaltReqTimeoutError: after 60 seconds. (Try 1 of 5)
2015-03-26 12:30:20,950 [salt.payload     ][INFO    ] SaltReqTimeoutError: after 60 seconds. (Try 2 of 5)
2015-03-26 12:30:28,715 [salt.minion      ][INFO    ] Authentication with master at [IP] successful!
2015-03-26 14:30:29,190 [salt.payload     ][INFO    ] SaltReqTimeoutError: after 7200 seconds. (Try 1 of 3)
2015-03-26 16:30:29,260 [salt.payload     ][INFO    ] SaltReqTimeoutError: after 7200 seconds. (Try 2 of 3)
2015-03-26 18:30:29,350 [salt.payload     ][INFO    ] SaltReqTimeoutError: after 7200 seconds. (Try 3 of 3)
2015-03-26 18:30:29,360 [salt.log.setup   ][ERROR   ] An un-handled exception was caught by salt's global exception handler:
SaltReqTimeoutError: SaltReqTimeoutError: after 21600 seconds, ran 3 tries
Traceback (most recent call last):
  File "/usr/bin/salt-minion", line 14, in <module>
    salt_minion()
  File "/usr/lib/python2.7/dist-packages/salt/scripts.py", line 57, in salt_minion
    minion.start()
  File "/usr/lib/python2.7/dist-packages/salt/__init__.py", line 262, in start
    self.prepare()
  File "/usr/lib/python2.7/dist-packages/salt/__init__.py", line 241, in prepare
    self.minion = salt.minion.Minion(self.config)
  File "/usr/lib/python2.7/dist-packages/salt/minion.py", line 604, in __init__
    opts['environment'],
  File "/usr/lib/python2.7/dist-packages/salt/pillar/__init__.py", line 91, in compile_pillar
    ret_pillar = self.sreq.crypted_transfer_decode_dictentry(load, dictkey='pillar', tries=3, timeout=7200)
  File "/usr/lib/python2.7/dist-packages/salt/transport/__init__.py", line 243, in crypted_transfer_decode_dictentry
    ret = self.sreq.send('aes', self.auth.crypticle.dumps(load), tries, timeout)
  File "/usr/lib/python2.7/dist-packages/salt/payload.py", line 224, in send
    'SaltReqTimeoutError: after {0} seconds, ran {1} tries'.format(timeout * tried, tried)
SaltReqTimeoutError: SaltReqTimeoutError: after 21600 seconds, ran 3 tries
```

As can be seen, the connection is lost, the minion restarts, it re-establishes the connection but (I assume) it is lost again immediately so the when it tries to get the pillars it times out and dies.

Looking at the default upstart salt-minion.conf (in ubuntu 14.04), we have the following:

```
# The respawn in the minion is known to cause problems
# because if the main minion process dies it has done
# so most likely for a good reason. Uncomment these
# two lines to enable respawn
#respawn
#respawn limit 10 5
```

As a result, the minion is dead.

**Comment by jfindlay** [comment]
@alf, what is the output of `salt --versions-report` and `salt minion test.versions_report` as necessary?  Thanks.

**Comment by alf** [comment]
Output from the master:

```
sudo salt --versions-report
                  Salt: 2014.7.2
                Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
                Jinja2: 2.7.2
              M2Crypto: 0.21.1
        msgpack-python: 0.3.0
          msgpack-pure: Not Installed
              pycrypto: 2.6.1
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 14.0.1
                  RAET: Not Installed
                   ZMQ: 4.0.4
                  Mako: 0.9.1
 Debian source package: 2014.7.2+ds-1trusty2
```

Output from the minion:

```
local:
                      Salt: 2014.7.2
                    Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
                    Jinja2: 2.7.2
                  M2Crypto: 0.21.1
            msgpack-python: 0.3.0
              msgpack-pure: Not Installed
                  pycrypto: 2.6.1
                   libnacl: Not Installed
                    PyYAML: 3.10
                     ioflo: Not Installed
                     PyZMQ: 14.0.1
                      RAET: Not Installed
                       ZMQ: 4.0.4
                      Mako: 0.9.1
     Debian source package: 2014.7.2+ds-1trusty2
```

**Comment by jfindlay** [comment]
@alf, do you have this problem on 2015.5.0 or later?  Thanks.

**Comment by alf** [comment]
From looking at the code, the exception is caught and handled in 2015.5.0 but not in 2014.7.6.

**Comment by jfindlay** [comment]
Since 2014.7 is no longer supported, I think this issue can be closed if you cannot reproduce it on 2015.5.0 or later.  Thanks @alf.
